compiler/lib/generate.ml

```diff
@@ -1218,64 +1224,68 @@ let rec translate_expr ctx queue loc in_tail_position e level : _ * J.statement_
   let cps = in_tail_position && Config.Flag.effects () in
   match e with
   | Apply { f; args; exact } ->
-      let args, prop, queue =
+      let (prop', f, deps), queue = access_queue queue f in
+      let args, prop, deps, queue =
         List.fold_right
-          ~f:(fun x (args, prop, queue) ->
-            let (prop', cx), queue = access_queue queue x in
-            cx :: args, or_p prop prop', queue)
+          ~f:(fun x (args, prop, deps, queue) ->
+            let (prop', cx, deps'), queue = access_queue queue x in
+            cx :: args, or_p prop prop', Var.Set.union deps deps', queue)
           args
-          ~init:([], mutator_p, queue)
+          ~init:([], mutator_p, deps, queue)
       in
-      let (prop', f), queue = access_queue queue f in
       let prop = or_p prop prop' in
       let e = apply_fun ctx f args exact cps loc in
-      (e, prop, queue), []
+      (e, prop, deps, queue), []
   | Block (tag, a, array_or_not) ->
-      let contents, prop, queue =
+      let contents, prop, deps, queue =
         List.fold_right
-          ~f:(fun x (args, prop, queue) ->
-            let (prop', cx), queue = access_queue queue x in
-            cx :: args, or_p prop prop', queue)
+          ~f:(fun x (args, prop, deps, queue) ->
+            let (prop', cx, deps'), queue = access_queue queue x in
+            cx :: args, or_p prop prop', Var.Set.union deps deps', queue)
           (Array.to_list a)
-          ~init:([], const_p, queue)
+          ~init:([], const_p, Code.Var.Set.empty, queue)
       in
       let x =
         match array_or_not with
         | Array -> Mlvalue.Array.make ~tag ~args:contents
         | NotArray | Unknown -> Mlvalue.Block.make ~tag ~args:contents
       in
-      (x, prop, queue), []
+      (x, prop, deps, queue), []
   | Field (x, n) ->
-      let (px, cx), queue = access_queue queue x in
-      (Mlvalue.Block.field cx n, or_p px mutable_p, queue), []
+      let (px, cx, deps), queue = access_queue queue x in
+      (Mlvalue.Block.field cx n, or_p px mutable_p, deps, queue), []
   | Closure (args, ((pc, _) as cont)) ->
       let loc = source_location ctx ~after:true pc in
       let clo = compile_closure ctx cont in
+      let deps = Code.Addr.Map.find pc ctx.free_vars in
       let clo =
         match clo with
         | (st, J.N) :: rem -> (st, J.U) :: rem
         | _ -> clo
       in
       let clo = J.EFun (None, List.map args ~f:(fun v -> J.V v), clo, loc) in
-      (clo, flush_p, queue), []
+      (clo, flush_p, deps, queue), []
```

Comment:
I think that since we have a `flush_p` here, no variable in `deps` will remain in the queue. So, we could just as well put `Var.Set.empty` here.

Reply:
I wish we could remove the `flush_p` here. We would need to know the free variables of the closure or rather the ones that are in the queue. But it's a bit tricky to compute that efficiently (`Freevar` only computes the free variables that are bound in a loop to avoid having to deal with large sets of variables.)

Reply:
Noted